file.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2016 Qubit Ltd.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	fileStartsCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "expexp_command_starts_total",
+			Help: "Counts of command starts",
+		},
+		[]string{"module"},
+	)
+	fileFailsCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "expexp_command_fails_total",
+			Help: "Count of commands with non-zero exits",
+		},
+		[]string{"module"},
+	)
+)
+
+func readFileWithDeadline(path string, t time.Time) ([]byte, time.Time, error) {
+	f, err := os.Open(path)
+	mtime := time.Time{}
+	if err != nil {
+		return nil, mtime, err
+	}
+	defer f.Close()
+	f.SetDeadline(t)
+
+	var size int
+	if info, err := f.Stat(); err == nil {
+		size64 := info.Size()
+		if int64(int(size64)) == size64 {
+			size = int(size64)
+		}
+		if info.Mode().IsRegular() {
+			mtime = info.ModTime()
+		}
+	}
+	size++ // one byte for final read at EOF
+
+	// If a file claims a small size, read at least 512 bytes.
+	// In particular, files in Linux's /proc claim size 0 but
```

Comment:
That part is shamelessly copied from golang source (function ReadFile)(https://cs.opensource.google/go/go/+/refs/tags/go1.20.2:src/os/file.go;l=674) I just added support for deadline. Maybe It will be better to state original source in the comment?

Reply:
Can we drop it and just go with ReadFile please. You are using this to hard code a 5 minute timeout. In reality the error here should be coming from a properly configured nfs/smb file system mount, which would work just as well with ReadFile.

Reply:
Ok SetReadDeadline is the right way to do this. I'm not native go programmer so I was not expecting asynchronous support on io.File interface.